File: iosApp/iosApp/Presentation/Main/CafeManagement/Schedule/ScheduleViewModel.swift
//
//  ScheduleViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/11.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class ScheduleViewModel: ObservableObject {
    private let castId: String?

    private let getScheduleManagementDataUseCase: GetScheduleManagementDataUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let updateCastScheduleUseCase: UpdateCastScheduleUseCase

    private let castEventPublisher: CastEventPublisher

    private let scheduleManagementEventPublisher: ScheduleManagementEventPublisher

    @Published private(set) var uiState = ScheduleUiState(isLoading: true)

    let event = PassthroughSubject<ScheduleEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func observeSession() {
        tasks[.session]?.cancel()
        tasks[.session] = Task {
            do {
                for try await _ in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.unbindCastEvent()
                    self.loadSchedule()
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func bindCastEvent(_ castId: String) {
        tasks[.castEvent]?.cancel()
        tasks[.castEvent] = Task {
            do {
                for try await event in asyncSequence(for: castEventPublisher.events) {
                    switch event {
                    case let event as Shared.CastEvent.Created:
                        if event.cast.id == castId {
                            self.loadSchedule()
                        }
                    case let event as Shared.CastEvent.Updated:
                        if event.cast.id == castId {
                            let cafeName = self.uiState.castSummary.subtitle.components(separatedBy: " / ").last ?? ""
                            let conceptRole: String
                            switch event.cast.conceptRole.lowercased() {
                            case "maid":
                                conceptRole = "schedule_concept_maid"
                            case "butler":
                                conceptRole = "schedule_concept_butler"
                            case "idol":
                                conceptRole = "schedule_concept_idol"
                            default:
                                conceptRole = event.cast.conceptRole.prefix(1).uppercased() + event.cast.conceptRole.dropFirst()
                            }
                            self.uiState.castSummary = ScheduleUiState.CastSummary(
                                title: event.cast.name,
                                subtitle: "\(conceptRole) / \(cafeName)",
                                badge: self.uiState.castSummary.badge,
                                initials: String(event.cast.name.prefix(2)).uppercased(),
                                profileImageUrl: event.cast.profileImage
                            )
                        }
                    case let event as Shared.CastEvent.Deleted:
                        if event.castId == castId {
                            self.loadSchedule()
                        }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func bindScheduleManagementEvent(_ castId: String) {
        tasks[.scheduleEvent]?.cancel()
        tasks[.scheduleEvent] = Task {
            do {
                for try await event in asyncSequence(for: scheduleManagementEventPublisher.events) {
                    if self.uiState.isSaving {
                        continue
                    }
                    switch event {
                    case let event as Shared.ScheduleManagementEvent.Updated:
                        if event.castId == castId {
                            self.loadSchedule(showLoading: false)
                            let message: String
                            switch event.status {
                            case .work:
                                message = "schedule_info_saved_work"
                            case .off:
                                message = "schedule_info_saved_off"
                            default:
                                message = "schedule_info_saved_vacation"
                            }
                            self.event.send(.showMessage(message))
                        }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func unbindCastEvent() {
        tasks.removeValue(forKey: .castEvent)?.cancel()
        tasks.removeValue(forKey: .scheduleEvent)?.cancel()
    }

    private func loadSchedule(showLoading: Bool = true) {
        tasks[.load]?.cancel()
        let currentPeriod = uiState.schedulePeriod
        let currentSelectedDayId = uiState.selectedDayId
        uiState.isLoading = showLoading
        uiState.isSaving = false
        uiState.errorMessage = nil
        uiState.infoMessage = nil
        uiState.pendingUpdates = []

        tasks[.load] = Task {
            do {
                let result = try await getScheduleManagementDataUseCase.invoke(castId: castId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let data = success.data as? Shared.ScheduleManagementData {
                    bindCastEvent(data.detail.cast.id)
                    bindScheduleManagementEvent(data.detail.cast.id)
                    let conceptRole: String
                    switch data.detail.cast.conceptRole.lowercased() {
                    case "maid":
                        conceptRole = "schedule_concept_maid"
                    case "butler":
                        conceptRole = "schedule_concept_butler"
                    case "idol":
                        conceptRole = "schedule_concept_idol"
                    default:
                        conceptRole = data.detail.cast.conceptRole.prefix(1).uppercased() + data.detail.cast.conceptRole.dropFirst()
                    }
                    uiState = ScheduleUiState(
                        managedCastId: data.detail.cast.id,
                        managedCafeId: data.detail.cafe.id,
                        managedCafeName: data.detail.cafe.name,
                        isLoading: false,
                        isSaving: false,
                        errorMessage: nil,
                        castSummary: .init(
                            title: data.detail.cast.name,
                            subtitle: "\(conceptRole) / \(data.detail.cafe.name)",
                            badge: "schedule_badge_cast_member",
                            initials: String(data.detail.cast.name.prefix(2)).uppercased(),
                            profileImageUrl: data.detail.cast.profileImage
                        ),
                        allWeekDays: data.weekDays,
                        allSchedules: data.daySchedules,
                        selectedDayId: currentSelectedDayId.isEmpty ? data.selectedDayId : currentSelectedDayId,
                        infoMessage: nil
                    ).applyingSchedulePeriod(currentPeriod)
                } else {
                    unbindCastEvent()
                    tasks.removeValue(forKey: .scheduleEvent)?.cancel()
                    uiState = ScheduleUiState(
                        isLoading: false,
                        isSaving: false,
                        errorMessage: "schedule_info_load_failed"
                    )
                }
            } catch {
                if Task.isCancelled { return }
                unbindCastEvent()
                tasks.removeValue(forKey: .scheduleEvent)?.cancel()
                uiState = ScheduleUiState(
                    isLoading: false,
                    isSaving: false,
                    errorMessage: "schedule_info_load_failed"
                )
            }
        }
    }

    func onAction(_ action: ScheduleAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .clickMore:
            uiState.infoMessage = "schedule_info_more_next_step"
        case .clickCalendar:
            let cafeId = uiState.managedCafeId
            let cafeName = uiState.managedCafeName
            if cafeId.isEmpty {
                uiState.infoMessage = "schedule_info_calendar_next_step"
            } else {
                event.send(.navigateToCastManagement(cafeId: cafeId, cafeName: cafeName))
            }
        case .selectPeriod(let period):
            uiState = uiState.applyingSchedulePeriod(period)
        case .selectDay(let id):
            uiState.selectedDayId = id
        case .clickEditDay(let id):
            guard let selected = uiState.schedules.first(where: { $0.id == id }) else { return }
            uiState.isEditSheetVisible = true
            uiState.editingScheduleId = selected.id
            uiState.editingScheduleTitle = selected.title
            uiState.editStatus = selected.status
            uiState.editStartTime = selected.timeLabel.components(separatedBy: " - ").first.flatMap { $0.contains(":") ? $0 : nil } ?? ScheduleUiState.defaultStartTime
            uiState.editEndTime = selected.timeLabel.components(separatedBy: " - ").last?.components(separatedBy: " ").first.flatMap { $0.contains(":") ? $0 : nil } ?? ScheduleUiState.defaultEndTime
            uiState.infoMessage = nil
        case .dismissEditSheet:
            uiState.isEditSheetVisible = false
            uiState.editingScheduleId = nil
        case .changeEditStatus(let status):
            uiState.editStatus = status
        case .changeEditStartTime(let value):
            uiState.editStartTime = value
        case .changeEditEndTime(let value):
            uiState.editEndTime = value
        case .submitEditDay:
            guard let editingId = uiState.editingScheduleId else { return }
            if uiState.editStatus == .work && ScheduleUiState.computeDurationMinutes(start: uiState.editStartTime, end: uiState.editEndTime) <= 0 {
                uiState.errorMessage = "schedule_error_end_after_start"
                return
            }
            let pendingUpdate = ScheduleUiState.PendingScheduleUpdate(
                date: editingId,
                status: uiState.editStatus,
                startTime: uiState.editStatus == .work ? uiState.editStartTime : nil,
                endTime: uiState.editStatus == .work ? uiState.editEndTime : nil
            )
            uiState.isEditSheetVisible = false
            uiState.editingScheduleId = nil
            uiState.errorMessage = nil
            uiState.infoMessage = "schedule_info_edit_applied"
            uiState.allSchedules = uiState.allSchedules.map { schedule in
                guard schedule.id == editingId else { return schedule }
                let timeLabel: String
                switch pendingUpdate.status {
                case .work:
                    timeLabel = "\(pendingUpdate.startTime ?? ScheduleUiState.defaultStartTime) - \(pendingUpdate.endTime ?? ScheduleUiState.defaultEndTime)"
                case .off:
                    timeLabel = "schedule_status_off"
                default:
                    timeLabel = "schedule_status_vacation"
                }
                let statusLabel: String
                switch pendingUpdate.status {
                case .work:
                    statusLabel = "schedule_status_work"
                case .off:
                    statusLabel = "schedule_status_off"
                default:
                    statusLabel = "schedule_status_vacation"
                }
                return ScheduleManagementDaySchedule(
                    id: pendingUpdate.date,
                    title: schedule.title,
                    timeLabel: timeLabel,
                    statusLabel: statusLabel,
                    isWorking: pendingUpdate.status == .work,
                    status: pendingUpdate.status
                )
            }
            uiState.allWeekDays = uiState.allWeekDays.map { day in
                guard day.id == editingId else { return day }
                return ScheduleManagementWeekDay(
                    id: day.id,
                    label: day.label,
                    number: day.number,
                    isWorking: pendingUpdate.status == .work
                )
            }
            uiState.pendingUpdates.removeAll { $0.date == editingId }
            uiState.pendingUpdates.append(pendingUpdate)
            uiState = uiState.applyingSchedulePeriod(uiState.schedulePeriod)
        case .clickSave:
            guard !uiState.managedCastId.isEmpty else { return }
            if uiState.pendingUpdates.isEmpty {
                uiState.infoMessage = "schedule_info_no_changes"
                uiState.errorMessage = nil
                return
            }
            uiState.isSaving = true
            uiState.errorMessage = nil
            uiState.infoMessage = nil

            let managedCastId = uiState.managedCastId
            let pendingUpdates = uiState.pendingUpdates
            let requestBatchId = "\(managedCastId)_\(Int(Date().timeIntervalSince1970 * 1000))"
            tasks[.submit]?.cancel()
            tasks[.submit] = Task { [weak self] in
                guard let self else { return }
                do {
                    for pendingUpdate in pendingUpdates {
                        let result = try await updateCastScheduleUseCase.invoke(
                            input: CastScheduleUpdate(
                                castId: managedCastId,
                                date: pendingUpdate.date,
                                status: pendingUpdate.status,
                                startTime: pendingUpdate.startTime,
                                endTime: pendingUpdate.endTime,
                                requestBatchId: requestBatchId
                            )
                        )
                        if Task.isCancelled { return }
                        if let failure = result as? AppResultFailure {
                            if let validation = failure.error as? AppErrorValidationFailed {
                                switch validation.reason {
                                case "start time is required":
                                    uiState.errorMessage = "schedule_error_start_required"
                                case "end time is required":
                                    uiState.errorMessage = "schedule_error_end_required"
                                case "end time must be after start time":
                                    uiState.errorMessage = "schedule_error_end_after_start"
                                default:
                                    uiState.errorMessage = "schedule_error_save_failed"
                                }
                            } else {
                                uiState.errorMessage = "schedule_error_week_save_failed"
                            }
                            uiState.isSaving = false
                            return
                        }
                    }
                    uiState.isSaving = false
                    uiState.errorMessage = nil
                    uiState.infoMessage = nil
                    uiState.pendingUpdates = []
                    self.event.send(.showMessage("schedule_event_week_saved"))
                } catch {
                    if Task.isCancelled { return }
                    uiState.isSaving = false
                    uiState.errorMessage = "schedule_error_week_save_failed"
                }
            }
        case .dismissInfoMessage:
            uiState.infoMessage = nil
            uiState.errorMessage = nil
        }
    }

    init(
        castId: String? = nil,
        getScheduleManagementDataUseCase: GetScheduleManagementDataUseCase = KoinInitializerKt.resolveGetScheduleManagementDataUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        updateCastScheduleUseCase: UpdateCastScheduleUseCase = KoinInitializerKt.resolveUpdateCastScheduleUseCase(),
        castEventPublisher: CastEventPublisher = KoinInitializerKt.resolveCastEventPublisher(),
        scheduleManagementEventPublisher: ScheduleManagementEventPublisher = KoinInitializerKt.resolveScheduleManagementEventPublisher()
    ) {
        self.castId = castId
        self.getScheduleManagementDataUseCase = getScheduleManagementDataUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.updateCastScheduleUseCase = updateCastScheduleUseCase
        self.castEventPublisher = castEventPublisher
        self.scheduleManagementEventPublisher = scheduleManagementEventPublisher

        observeSession()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case session
        case castEvent
        case scheduleEvent
        case load
        case submit
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/schedule/ScheduleViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.schedule

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppError
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.publisher.CastEventPublisher
import com.hhp227.concafe.domain.event.publisher.ScheduleManagementEventPublisher
import com.hhp227.concafe.domain.model.CastScheduleStatus
import com.hhp227.concafe.domain.model.CastScheduleUpdate
import com.hhp227.concafe.domain.model.ScheduleManagementDaySchedule
import com.hhp227.concafe.domain.usecase.GetScheduleManagementDataUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import com.hhp227.concafe.domain.usecase.UpdateCastScheduleUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.event.CastEvent as CastDomainEvent
import com.hhp227.concafe.domain.event.ScheduleManagementEvent as ScheduleManagementDomainEvent

class ScheduleViewModel(
    private val castId: String? = null,
    private val getScheduleManagementDataUseCase: GetScheduleManagementDataUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val updateCastScheduleUseCase: UpdateCastScheduleUseCase,
    private val castEventPublisher: CastEventPublisher,
    private val scheduleManagementEventPublisher: ScheduleManagementEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(ScheduleUiState(isLoading = true))
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<ScheduleEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun observeSession() {
        jobs[TaskKey.OBSERVE_SESSION]?.cancel()
        jobs[TaskKey.OBSERVE_SESSION] = viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest {
                unbindCastEvent()
                loadSchedule()
            }
        }
    }

    private fun bindCastEvent(castId: String) {
        jobs[TaskKey.OBSERVE_CAST_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAST_EVENT] = viewModelScope.launch {
            castEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CastDomainEvent.Created -> if (event.cast.id == castId) {
                        loadSchedule()
                    }
                    is CastDomainEvent.Updated -> if (event.cast.id == castId) {
                        _uiState.update { state ->
                            val cafeName = state.castSummary.subtitle.substringAfter(" / ", "")
                            val conceptRole = when (event.cast.conceptRole.lowercase()) {
                                "maid" -> SCHEDULE_CONCEPT_MAID
                                "butler" -> SCHEDULE_CONCEPT_BUTLER
                                "idol" -> SCHEDULE_CONCEPT_IDOL
                                else -> event.cast.conceptRole.replaceFirstChar { char -> char.uppercase() }
                            }
                            state.copy(
                                castSummary = state.castSummary.copy(
                                    title = event.cast.name,
                                    subtitle = "$conceptRole / $cafeName",
                                    initials = event.cast.name.take(2).uppercase(),
                                    profileImageUrl = event.cast.profileImage
                                )
                            )
                        }
                    }
                    is CastDomainEvent.Deleted -> if (event.castId == castId) {
                        loadSchedule()
                    }
                }
            }
        }
    }

    private fun bindScheduleManagementEvent(castId: String) {
        jobs[TaskKey.OBSERVE_SCHEDULE_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_SCHEDULE_EVENT] = viewModelScope.launch {
            scheduleManagementEventPublisher.events.collectLatest { event ->
                when (event) {
                    is ScheduleManagementDomainEvent.Updated -> if (event.castId == castId) {
                        if (_uiState.value.isSaving) {
                            return@collectLatest
                        }
                        loadSchedule(showLoading = false)
                        _event.emit(
                            ScheduleEvent.ShowMessage(
                                when (event.status) {
                                    CastScheduleStatus.WORK -> "schedule_info_saved_work"
                                    CastScheduleStatus.OFF -> "schedule_info_saved_off"
                                    CastScheduleStatus.VACATION -> "schedule_info_saved_vacation"
                                }
                            )
                        )
                    }
                }
            }
        }
    }

    private fun unbindCastEvent() {
        jobs.remove(TaskKey.OBSERVE_CAST_EVENT)?.cancel()
        jobs.remove(TaskKey.OBSERVE_SCHEDULE_EVENT)?.cancel()
    }

    private fun loadSchedule(showLoading: Boolean = true) {
        val currentPeriod = _uiState.value.schedulePeriod
        val currentSelectedDayId = _uiState.value.selectedDayId

        _uiState.update {
            it.copy(
                isLoading = showLoading,
                isSaving = false,
                errorMessage = null,
                infoMessage = null,
                pendingUpdates = emptyList()
            )
        }
        viewModelScope.launch {
            when (val result = getScheduleManagementDataUseCase.invoke(castId)) {
                is AppResult.Success -> {
                    bindCastEvent(result.data.detail.cast.id)
                    bindScheduleManagementEvent(result.data.detail.cast.id)
                    val data = result.data
                    val conceptRole = when (data.detail.cast.conceptRole.lowercase()) {
                        "maid" -> SCHEDULE_CONCEPT_MAID
                        "butler" -> SCHEDULE_CONCEPT_BUTLER
                        "idol" -> SCHEDULE_CONCEPT_IDOL
                        else -> data.detail.cast.conceptRole.replaceFirstChar { char -> char.uppercase() }
                    }
                    _uiState.value = ScheduleUiState(
                        managedCastId = data.detail.cast.id,
                        managedCafeId = data.detail.cafe.id,
                        managedCafeName = data.detail.cafe.name,
                        isLoading = false,
                        isSaving = false,
                        errorMessage = null,
                        castSummary = ScheduleUiState.CastSummary(
                            title = data.detail.cast.name,
                            subtitle = "$conceptRole / ${data.detail.cafe.name}",
                            badge = "schedule_badge_cast_member",
                            initials = data.detail.cast.name.take(2).uppercase(),
                            profileImageUrl = data.detail.cast.profileImage
                        ),
                        allWeekDays = data.weekDays,
                        allSchedules = data.daySchedules,
                        selectedDayId = currentSelectedDayId.ifBlank { data.selectedDayId },
                        infoMessage = null
                    ).withSchedulePeriod(currentPeriod)
                }
                is AppResult.Failure -> {
                    unbindCastEvent()
                    jobs.remove(TaskKey.OBSERVE_SCHEDULE_EVENT)?.cancel()
                    _uiState.value = ScheduleUiState(
                        isLoading = false,
                        errorMessage = "schedule_info_load_failed",
                        infoMessage = null
                    )
                }
            }
        }
    }

    fun onAction(action: ScheduleAction) {
        when (action) {
            ScheduleAction.ClickBack -> {
                viewModelScope.launch {
                    _event.emit(ScheduleEvent.NavigateBack)
                }
            }
            ScheduleAction.ClickCalendar -> {
                val state = _uiState.value
                val cafeId = state.managedCafeId
                val cafeName = state.managedCafeName
                if (cafeId.isBlank()) {
                    _uiState.update { it.copy(infoMessage = "schedule_info_calendar_next_step") }
                } else {
                    viewModelScope.launch {
                        _event.emit(ScheduleEvent.NavigateToCastManagement(cafeId, cafeName))
                    }
                }
            }
            is ScheduleAction.SelectPeriod -> {
                _uiState.update { state -> state.withSchedulePeriod(action.period) }
            }
            is ScheduleAction.SelectDay -> {
                _uiState.update { state ->
                    state.copy(
                        selectedDayId = action.dayId
                    )
                }
            }
            is ScheduleAction.ClickEditDay -> {
                val selected = _uiState.value.schedules.firstOrNull { it.id == action.dayId } ?: return
                _uiState.update {
                    it.copy(
                        isEditSheetVisible = true,
                        editingScheduleId = selected.id,
                        editingScheduleTitle = selected.title,
                        editStatus = selected.status,
                        editStartTime = selected.timeLabel.substringBefore(" - ").takeIf { time -> ":" in time } ?: ScheduleUiState.DEFAULT_START_TIME,
                        editEndTime = selected.timeLabel.substringAfter(" - ", ScheduleUiState.DEFAULT_END_TIME).substringBefore(" ").takeIf { time -> ":" in time } ?: ScheduleUiState.DEFAULT_END_TIME,
                        infoMessage = null
                    )
                }
            }
            ScheduleAction.DismissEditSheet -> {
                _uiState.update {
                    it.copy(
                        isEditSheetVisible = false,
                        editingScheduleId = null
                    )
                }
            }
            is ScheduleAction.ChangeEditStatus -> {
                _uiState.update { it.copy(editStatus = action.status) }
            }
            is ScheduleAction.ChangeEditStartTime -> {
                _uiState.update { it.copy(editStartTime = action.value) }
            }
            is ScheduleAction.ChangeEditEndTime -> {
                _uiState.update { it.copy(editEndTime = action.value) }
            }
            ScheduleAction.SubmitEditDay -> {
                val currentState = _uiState.value
                val editingId = currentState.editingScheduleId ?: return
                if (currentState.editStatus == CastScheduleStatus.WORK && currentState.editStartTime >= currentState.editEndTime) {
                    _uiState.update { it.copy(errorMessage = "schedule_error_end_after_start") }
                    return
                }
                _uiState.update { state ->
                    val pendingUpdate = ScheduleUiState.PendingScheduleUpdate(
                        date = editingId,
                        status = currentState.editStatus,
                        startTime = currentState.editStartTime.takeIf { currentState.editStatus == CastScheduleStatus.WORK },
                        endTime = currentState.editEndTime.takeIf { currentState.editStatus == CastScheduleStatus.WORK }
                    )
                    state.copy(
                        isEditSheetVisible = false,
                        editingScheduleId = null,
                        errorMessage = null,
                        infoMessage = "schedule_info_edit_applied",
                        allSchedules = state.allSchedules.map { schedule ->
                            if (schedule.id == editingId) {
                                val isWorking = pendingUpdate.status == CastScheduleStatus.WORK
                                val timeLabel = when (pendingUpdate.status) {
                                    CastScheduleStatus.WORK -> "${pendingUpdate.startTime ?: ScheduleUiState.DEFAULT_START_TIME} - ${pendingUpdate.endTime ?: ScheduleUiState.DEFAULT_END_TIME}"
                                    CastScheduleStatus.OFF -> "schedule_status_off"
                                    CastScheduleStatus.VACATION -> "schedule_status_vacation"
                                }
                                val statusLabel = when (pendingUpdate.status) {
                                    CastScheduleStatus.WORK -> "schedule_status_work"
                                    CastScheduleStatus.OFF -> "schedule_status_off"
                                    CastScheduleStatus.VACATION -> "schedule_status_vacation"
                                }
                                ScheduleManagementDaySchedule(
                                    id = pendingUpdate.date,
                                    title = schedule.title,
                                    timeLabel = timeLabel,
                                    statusLabel = statusLabel,
                                    isWorking = isWorking,
                                    status = pendingUpdate.status
                                )
                            } else {
                                schedule
                            }
                        },
                        allWeekDays = state.allWeekDays.map { day ->
                            if (day.id == editingId) {
                                day.copy(isWorking = pendingUpdate.status == CastScheduleStatus.WORK)
                            } else {
                                day
                            }
                        },
                        pendingUpdates = state.pendingUpdates
                            .filterNot { it.date == editingId } + pendingUpdate
                    ).withSchedulePeriod(state.schedulePeriod)
                }
            }
            ScheduleAction.ClickMore -> {
                _uiState.update { it.copy(infoMessage = "schedule_info_more_next_step") }
            }
            ScheduleAction.ClickSave -> {
                val currentState = _uiState.value
                val managedCastId = currentState.managedCastId.ifBlank { return }
                if (currentState.pendingUpdates.isEmpty()) {
                    _uiState.update { it.copy(infoMessage = "schedule_info_no_changes", errorMessage = null) }
                    return
                }
                _uiState.update { it.copy(isSaving = true, errorMessage = null, infoMessage = null) }
                jobs[TaskKey.SUBMIT]?.cancel()
                jobs[TaskKey.SUBMIT] = viewModelScope.launch {
                    val pendingUpdates = currentState.pendingUpdates
                    for (pendingUpdate in pendingUpdates) {
                        when (
                            val result = updateCastScheduleUseCase.invoke(
                                CastScheduleUpdate(
                                    castId = managedCastId,
                                    date = pendingUpdate.date,
                                    status = pendingUpdate.status,
                                    startTime = pendingUpdate.startTime,
                                    endTime = pendingUpdate.endTime
                                )
                            )
                        ) {
                            is AppResult.Success -> Unit
                            is AppResult.Failure -> {
                                _uiState.update {
                                    val error = result.error
                                    it.copy(
                                        isSaving = false,
                                        errorMessage = when (error) {
                                            is AppError.ValidationFailed -> when (error.reason) {
                                                "start time is required" -> "schedule_error_start_required"
                                                "end time is required" -> "schedule_error_end_required"
                                                "end time must be after start time" -> "schedule_error_end_after_start"
                                                else -> "schedule_error_save_failed"
                                            }
                                            else -> "schedule_error_week_save_failed"
                                        }
                                    )
                                }
                                return@launch
                            }
                        }
                    }
                    _uiState.update {
                        it.copy(
                            isSaving = false,
                            errorMessage = null,
                            infoMessage = null,
                            pendingUpdates = emptyList()
                        )
                    }
                    _event.emit(ScheduleEvent.ShowMessage("schedule_event_week_saved"))
                }
            }
            ScheduleAction.DismissInfoMessage -> {
                _uiState.update { it.copy(infoMessage = null, errorMessage = null) }
            }
        }
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    init {
        observeSession()
    }

    companion object {
        private const val SCHEDULE_CONCEPT_MAID = "schedule_concept_maid"
        private const val SCHEDULE_CONCEPT_BUTLER = "schedule_concept_butler"
        private const val SCHEDULE_CONCEPT_IDOL = "schedule_concept_idol"
    }
}

private enum class TaskKey {
    OBSERVE_SESSION,
    OBSERVE_CAST_EVENT,
    OBSERVE_SCHEDULE_EVENT,
    SUBMIT
}
